### PR TITLE
(Closes #386) Extend reader to allow encoding info. to be ignored.

### LINKED ISSUE #386: Support extension to permit in-line '!' comments in fixed-format Fortran
The OpenMPI Fortran include files are fixed format (they have the `-*- fortran -*-` header and non-comment lines begin with 6 spaces). However, they also have some 'inline' style comments, e.g.:

    ! MPI F08 conformance
    !
          logical MPI_SUBARRAYS_SUPPORTED
          logical MPI_ASYNC_PROTECTS_NONBLOCKING
          ! Hard-coded for .false. for now
          parameter (MPI_SUBARRAYS_SUPPORTED= .false.)

and fparser does not accept such comments giving rise to a syntax error. Given that this is in OpenMPI code this must mean that the vast majority of compilers have no problem, despite it being non-standards compliant F77. I'm therefore inclined to say that fparser should just silently permit such comments too.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
This is ready for a look now. Probably one for @hiker as @sergisiso has the in-lining to play with. It's not a big change but while doing it I realised that we weren't able to handle inline comments on an 'include' line so I've fixed that too.
It is however a change in the default behaviour of the parser. I think it's a change in the right direction but I'm open to other opinions :-)

File: src/fparser/common/tests/test_readfortran.py [hunk 1]
```diff
@@ -858,7 +858,28 @@ def test_string_reader():
...
+        # File location with full path.
+        rinput = str(sfile.absolute())
+    else:
+        rinput = source
+    reader = reader_cls(rinput)
+    # By default the encoding information is ignored so the format should be
+    # free format, not strict.
+    assert reader.format == FortranFormat(True, False)
+    # Now test when the reader takes notice of the encoding information.
+    reader2 = reader_cls(rinput, ignore_encoding=False)
```
Comment (hiker) [inline src/fparser/common/tests/test_readfortran.py:880]: Maybe I am overdoing it, but just in case maybe add a test for `ignore_ending=True` here.

**Review by hiker — CHANGES_REQUESTED** [review]
Great job. All tests work for me. I did not run the usual PSyclone style checking, since I got 100's of error messages, clearly not caused by this PR.

I had two or three minor comments only.

File: doc/conf.py [hunk 2]
```diff
@@ -95,6 +95,7 @@
 # ones.
 extensions = [
     'sphinx.ext.autodoc',
+    'sphinx.ext.autosummary',
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/conf.py:98]: I thought I'd found that I needed it but I was getting confused with another branch where I'm adding a reference guide. Reverted now.

File: src/fparser/common/readfortran.py [hunk 3]
```diff
@@ -791,22 +791,13 @@ def next(self, ignore_comments=None):
...
-                        # file. Setting reader to None indicates that
-                        # we should now read from the main reader.
-                        self.reader = None
+                    # Return a line from the include.
+                    return self.reader.next(ignore_comments)
+                except StopIteration:
+                    # There is nothing left in the include
+                    # file. Setting reader to None indicates that
+                    # we should now read from the main reader.
+                    self.reader = None
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/common/readfortran.py:800]: Yes, I couldn't understand why the code was like it was as `next()` is already designed to be called recursively...so, I tried it and it worked :-)

File: src/fparser/common/sourceinfo.py [hunk 4]
```diff
@@ -227,25 +227,35 @@ def mode(self):
...
-        return FortranFormat(False, True)
-    if _HAS_FIX_HEADER(firstline):
-        return FortranFormat(False, False)
-    if _HAS_FREE_HEADER(firstline):
-        return FortranFormat(True, False)
-    if _HAS_PYF_HEADER(firstline):
-        return FortranFormat(True, True)
+    if not ignore_encoding:
+        # We check to see whether the file contains a comment describing its
+        # encoding. This is nothing to do with the Fortran standard (see e.g.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/common/sourceinfo.py:247]: Thanks.

File: src/fparser/common/tests/test_readfortran.py (hunk 1 above)
Comment (arporter) [reply src/fparser/common/tests/test_readfortran.py:880]: You can never have too many tests (within reason).

**Comment by arporter** [comment]
Ready for another look now.

File: doc/conf.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hiker) [reply doc/conf.py]: Thanks.

File: src/fparser/common/tests/test_readfortran.py (hunk 1 above)
Comment (hiker) [reply src/fparser/common/tests/test_readfortran.py:883]: I hope that was within reason .. thanks!

**Review by hiker — APPROVED** [review]
All issues have been addressed, processing to merge after CI.

**Pull request merged**